stdgo/reflect/Reflect.hx

```diff
@@ -416,30 +435,111 @@ class _GoType__extension {
 		}
 		return false;
 	}
+
+	public static function comparable(gt:Type):Bool {
+		switch (gt) {
+			case GT_slice(_), GT_func(_, _), GT_map(_, _):
+				return false;
+			case GT_array(elem, _):
+				return elem.comparable();
+			case GT_struct(fields):
+				for (fld in fields)
+					if (!fld.type.comparable())
+						return false;
+				return true;
+			case _:
+				return true;
+		}
+	}
+}
+
+function unserializeType(s:String):Type {
+	var unserializer = new Unserializer(s);
+	var ret = unserializer.unserialize();
+	if (Reflect.isEnumValue(ret)) {
+		return ret;
+	}
+	return GT_invalid;
 }
 
-function ptrTo(t:GoType):GoType {
+function ptrTo(t:Type):Type {
 	return GT_ptr(t);
 }
 
-function typeOf(iface:AnyInterface):GoType {
+var namedTypeMap:Map<GoString, NamedTypeData> = [];
+
+function typeOf(iface:AnyInterface):Type {
+	if (iface == null)
+		return GT_unsafePointer;
 	var tn = iface.typeName();
 	var v = iface.value();
-	if (tn == "*unknown") { // TODO move similar code to Pointer.hx
-		var elem = (v : Dynamic).get();
-		var elemT = typeOfAnyHaxe(elem);
-		if (elemT != GT_invalid)
-			return GT_ptr(elemT);
+	switch (tn) {
+		case "*unknown": // TODO this code may be in the wrong place
+			var elem = (v : Dynamic).get();
+			var elemT = typeOfAnyHaxe(elem);
+			if (elemT != GT_invalid)
+				return GT_ptr(elemT);
+		case _:
 	}
-	var gt = makeUnNamedGoType(tn);
+	var gt = makeTypeFromTypeName(tn);
 	if (gt != GT_invalid)
 		return gt;
 	return typeOfAnyHaxe(v);
 }
 
-function typeOfAnyHaxe(v:Dynamic):GoType {
+function makeTypeFromTypeName(typeName:String):Type {
+	// trace("mNBT:", typeName);
+
+	if (StringTools.startsWith(typeName, "wy21:stdgo.reflect.GT_enum"))
+		return unserializeType(typeName);
+
+	var pointerTo = false;
+	if (typeName.length == 0)
+		return new Type(GT_invalid);
+	if (StringTools.startsWith(typeName, "*")) {
+		pointerTo = true;
+		typeName = typeName.substr(1);
+	}
+
+	switch (typeName) {
+		case "invalid", "array", "chan", "func", "map", "ptr", "slice", "struct", "namedType":
```

Comment:
Won't these typeNames get missed as for instance a slice int as a typeName will be "[]int"? Let me know if this shouldn't be the case, because from what I understood typeNames were also going to be used extensively by type-switches as well.

Reply:
The types that can be created from a text value are those that are:
- the xxxxx of a GT_xxxxx enum that has no parameters
- the name of a Haxe class (for interfaces and named types)
- the [Haxe serialisation format](https://haxe.org/manual/std-serialization-format.html) encoded form of a GT_enum (allowing full type information to be passed using the typeName)

Other type information is deduced by inspection, as prompted by the typeName list you highlight. 

I envisage that the type switch code will not use text comparison, but will be a series of tests of the form `switchType.Implements(caseType)`, where the type values are `reflect.Type`.

Reply:
Ah gotcha sounds good, thank you for the explanation, makes much more sense now.